components/raftstore/src/store/peer.rs

```diff
@@ -3496,12 +3509,42 @@ where
         Ok((min_m, min_c))
     }
 
-    fn pre_propose_prepare_merge<T>(
-        &self,
+    fn pre_propose_prepare_merge<T: Transport>(
+        &mut self,
         ctx: &mut PollContext<EK, ER, T>,
         req: &mut RaftCmdRequest,
     ) -> Result<()> {
         let last_index = self.raft_group.raft.raft_log.last_index();
+
+        // Record current proposed index. If there are some in-memory pessimistic locks, we should
+        // wait until applying to the proposed index before proposing pessimistic locks and
+        // PrepareMerge. Otherwise, if an already proposed command will remove a pessimistic lock,
+        // we will make some deleted locks appear again.
+        {
```

Comment:
Is it a good idea to move this check after checking `merge_max_log_gap`, admin requests and `raft_entry_max_size`? And when there is a `prepare_merge_fence`, only check apply index and do not check these again.

Reply:
> Is it a good idea to move this check after checking `merge_max_log_gap`, admin requests and `raft_entry_max_size`?

Ah, yes. We shouldn't block write if the log gap is too large.

Now I move it after these checks.

> And when there is a `prepare_merge_fence`, only check apply index and do not check these again.

I'm not sure... It may make the code more complex. And when the applied index reaches the fence, some pessimistic locks may be deleted and the log gap may shrink too. It will be more likely to pass the size check. Considering the frequency of region merge, perhaps a second calculation is acceptable.

Reply:
> And when the applied index reaches the fence, some pessimistic locks may be deleted and the log gap may shrink too. It will be more likely to pass the size check.

The size check can't block merge in this case, but is it possible there is a new admin request within the log gap that forbid to merge because admin requests can pass `prepare_merge_fence` check? We should clear fence if it happens.

Reply:
Clear the fence when:
1. pass all checks and start proposing `PrepareMerge`.
2. fail to propose `PrepareMerge` and the fence exsits.

Reply:
> is it possible there is a new admin request within the log gap that forbid to merge because admin requests can pass `prepare_merge_fence` check

Definitely. Now, I use the similar way as before, clear the fence before prepare merge checks. Hope to reduce subtleties burning brains...